chore: fix header generation on macos

jsantell opened the pull request:
Provide a script to generate headers that uses codegen_units=1 when building on macos to circumvent a header generation issue. Update CI and C makefile to use this script. Fixes #473.
Also fixes a linker issue on macos for the C example. Fixes #476.

justinabrahms commented on scripts/xcode_local_build:
```diff
@@ -0,0 +1,53 @@
+#!/usr/bin/env bash
+set -e
+set -x
```
```suggestion
set -euxo pipefail
```

So you're missing `-u` which complains about unused variables (useful for dumb mistakes) and `-o pipefail` which says "if I use a pipe and it errors, then error".  Otherwise, you can have situations like:  

```sh
$ exit 1 | echo 'works'
works
```
when you really want it to error/exit.

jsantell replied:
ty for the feedback! I'll apply these throughout

justinabrahms commented on scripts/xcode_local_build:
```diff
@@ -0,0 +1,53 @@
+SCRIPT_DIR=$( cd -- "$( dirname -- "${BASH_SOURCE[0]}" )" &> /dev/null && pwd )
+PROJECT_DIR="$SCRIPT_DIR/../"
+
+pushd $PROJECT_DIR
```
```suggestion
cd $PROJECT_DIR
```

justinabrahms commented on scripts/xcode_local_build:
```diff
@@ -0,0 +1,53 @@
+    # cargo build --package noosphert buie --target $TARGET --locked
+done
+
+mkdir -p ./target/macos
+mkdir -p ./target/simulator
```
```suggestion
mkdir -p ./target/{macos,simulator}
```

justinabrahms commented on scripts/xcode_local_build:
```diff
@@ -0,0 +1,53 @@
+for TARGET in "${TARGETS[@]}"; do
+    rustup target install $TARGET
+    cargo build --package noosphere --release --target $TARGET --locked
+    # cargo build --package noosphert buie --target $TARGET --locked
```
I don't know what buie is, but maybe remove this?
```suggestion
```

justinabrahms commented on scripts/xcode_local_build:
```diff
@@ -0,0 +1,53 @@
+    -library ./target/aarch64-apple-ios/release/libnoosphere.a \
+    -headers ./target/headers/include/ \
+    -output ./LibNoosphere.xcframework
+
+popd
+
+set +x
+set +e
```
```suggestion
```
You don't need to leave the directory or unset your flags. Bash scripts run as a subprocess and don't affect the outer scope. This would only be important if you were writing some library that gets sourced (which runs in your currents scope). This shouldn't be that.

justinabrahms reviewed (commented):
A few suggestions around bash scripts.

jsantell commented:
Cleaned up the bash script from @justinabrahms' feedback; added a fix for #476 to link to `-framework Security` when running the C example on macos due to our `ring` dependencies.

jsantell commented on scripts/generate_headers:
```diff
@@ -0,0 +1,21 @@
+if [[ "$OSTYPE" == "darwin"* ]]; then
+	# macos linker fails to generate all exports without this flag
+	# https://github.com/subconsciousnetwork/noosphere/issues/473
+	CARGO_PROFILE_DEV_CODEGEN_UNITS=1 cargo run --verbose --package noosphere --example generate_header --features headers --locked
```
@cdata this can be set via `[profile.dev]` for macos only, but don't believe we can target only the header gen example.

cdata replied:
What are the side-effects? Is it just a perf difference?

jsantell replied:
prevents LTO maybe? https://doc.rust-lang.org/cargo/reference/profiles.html#codegen-units

jsantell replied:
serendipitous timing https://nnethercote.github.io/2023/07/11/back-end-parallelism-in-the-rust-compiler.html

cdata replied:
Let's just stick to the script as written. We can revisit if it seems unsatisfying down the road, and I would rather not compromise all builds for the sake of this problem.

jsantell replied:
Aye definitely not something we want on all builds, this just runs on macos's header-gen binary

jsantell replied:
This gets the job done, but there's a side effect of needing to rebuild more often as this flag invalidates debug builds (that are missing this flag)

cdata commented:
Conventionally we use `.sh` extensions for bash scripts. I don't think this is strictly necessary, but I would like to hear an argument for changing the convention before we commit to it.

I only have one argument in favor of `.sh`, which is that file extensions are valuable for inferring the file type without reading its contents (a topic top of mind for me as I work on `orb` file rendering).

jsantell commented:
Updated, adding `swift-test` to run swift tests locally, as well as adding some flags to manage debug/release builds. Consistent flag handling and usage docs added where needed.

These 3 scripts are consistent with themselves, but we need to make a few decisions on conventions; lets hammer out what those are and update here, or after, accordingly. What I'm using here are weak personal preferences:

* Indentation: 2 spaces
* Shebang: `#!/usr/bin/env bash`
* Path normalization:  Prefer `$( cd -- "$( dirname -- "${BASH_SOURCE[0]}" )" &> /dev/null && pwd )` over `git rev-parse --show-toplevel`
* Naming: Use `-` (more bashy, easier to type) and omit extension (shebang clarifies any ambiguities if needed, will be executed more than read, and nicer when setting scripts in $PATH

jsantell commented on scripts/generate-headers:
```diff
@@ -0,0 +1,57 @@
+PROJECT_DIR="$SCRIPT_DIR/../"
+cd $PROJECT_DIR
+
+DEFAULT_OUT="`realpath $PROJECT_DIR`/target/headers"
```
`realpath` needs `brew install coreutils` on Macos, and other solutions like `readlink -f` don't work on macos/BSD -- this is only really needed since it's displayed in the `usage`. do you have any cross-platform preferences for something like this? @justinabrahms

jsantell commented on scripts/swift-test:
```diff
@@ -0,0 +1,62 @@
+cp Package.swift $PLACEHOLDER
+
+FRAMEWORK_PATH="./target/framework/$PROFILE/LibNoosphere.xcframework"
+echo $FRAMEWORK_PATH
```
TODO: Remove

jsantell commented on scripts/build-framework:
```diff
@@ -0,0 +1,101 @@
+PROFILE="debug"
+
+function usage() {
+  echo "build-framework: a utility to build an Apple XCFramework for Noosphere."
```
Just copying http://docopt.org/ styling here

cdata commented on scripts/build-framework.sh:
```diff
@@ -0,0 +1,101 @@
+RELEASE_FLAG=""
+PROFILE="debug"
+
+function usage() {
```
:fire:

cdata reviewed (approved):
This is awesome, and clears up some huge FUD I had around header generation. Thanks @jsantell
